pkg/nginxprocess/process.go

```diff
@@ -0,0 +1,142 @@
+	PID     int32
+	PPID    int32 // parent PID
+}
+
+// IsWorker returns true if the process is a NGINX worker process.
+func (p *Process) IsWorker() bool { return strings.Contains(p.Cmd, "worker") }
+
+// IsShuttingDown returns true if the process is shutting down. This can identify workers that are in the process of a
+// graceful shutdown. See [changing NGINX configuration] for more details.
+//
+// [changing NGINX configuration]: https://nginx.org/en/docs/control.html#reconfiguration
+func (p *Process) IsShuttingDown() bool { return strings.Contains(p.Cmd, "is shutting down") }
+
+// IsHealthy uses Status flags to judge process health. Only works on processes created using [WithStatus].
+func (p *Process) IsHealthy() bool {
+	return p.Status != "" && !strings.Contains(p.Status, process.Zombie)
+}
+
+type options struct {
+	loadStatus bool
+}
+
+// Option customizes how processes are gathered from the OS.
+type Option interface{ apply(opts *options) }
+
+type optionFunc func(*options)
+
+func (f optionFunc) apply(o *options) { f(o) }
+
+// WithStatus runs an additional lookup to load the process status.
+func WithStatus(v bool) Option { //nolint:ireturn // functional options can be opaque
+	return optionFunc(func(o *options) { o.loadStatus = v })
+}
+
+func convert(ctx context.Context, p *process.Process, o options) (*Process, error) {
+	if ctx.Err() != nil { // fail fast if we've canceled
+		return nil, ctx.Err()
+	}
+
+	name, _ := p.NameWithContext(ctx) // slow: shells out to ps
+	if name != "nginx" {
+		return nil, errNotAnNginxProcess
+	}
+
+	cmdLine, _ := p.CmdlineWithContext(ctx) // slow: shells out to ps
+	// ignore nginx processes in the middle of an upgrade
+	if !strings.HasPrefix(cmdLine, "nginx:") || strings.Contains(cmdLine, "upgrade") {
+		return nil, errNotAnNginxProcess
+	}
+
+	var status string
+	if o.loadStatus {
+		flags, _ := p.StatusWithContext(ctx) // slow: shells out to ps
+		status = strings.Join(flags, " ")
+	}
+
+	// unconditionally run fast lookups
+	var created time.Time
+	if millisSinceEpoch, err := p.CreateTimeWithContext(ctx); err == nil {
+		nanos := millisSinceEpoch * int64(time.Millisecond)
+		created = time.Unix(0, nanos)
+	}
+	ppid, _ := p.PpidWithContext(ctx)
+	exe, _ := p.ExeWithContext(ctx)
+
+	return &Process{
+		PID:     p.Pid,
+		PPID:    ppid,
+		Name:    name,
+		Cmd:     cmdLine,
+		Created: created,
+		Status:  status,
+		Exe:     exe,
+	}, ctx.Err()
+}
+
+// List returns a slice of all NGINX processes. Returns a zero-length slice if no NGINX processes are found.
+func List(ctx context.Context, opts ...Option) (ret []*Process, err error) {
+	o := options{}
+	for _, opt := range opts {
+		opt.apply(&o)
+	}
+	processes, err := process.ProcessesWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, p := range processes {
+		pr, cerr := convert(ctx, p, o)
+		if errors.Is(cerr, errNotAnNginxProcess) {
```

Comment:
use the new func for consistency? `IsNotNginxErr`

Reply:
done